org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/GradlePreferenceChangeListener.java

```diff
@@ -0,0 +1,74 @@
+ *     Red Hat Inc. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.ls.core.internal.managers;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+import org.eclipse.buildship.core.GradleDistribution;
+import org.eclipse.buildship.core.WrapperGradleDistribution;
+import org.eclipse.buildship.core.internal.CorePlugin;
+import org.eclipse.buildship.core.internal.configuration.ProjectConfiguration;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.ProjectUtils;
+import org.eclipse.jdt.ls.core.internal.preferences.IPreferencesChangeListener;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
+import org.eclipse.jdt.ls.core.internal.preferences.Preferences;
+import org.eclipse.jdt.ls.internal.gradle.checksums.ValidationResult;
+import org.eclipse.jdt.ls.internal.gradle.checksums.WrapperValidator;
+
+/**
+ * @author snjeza
+ *
+ */
+public class GradlePreferenceChangeListener implements IPreferencesChangeListener {
+	@Override
+	public void preferencesChange(Preferences oldPreferences, Preferences newPreferences) {
+		ProjectsManager projectsManager = JavaLanguageServerPlugin.getProjectsManager();
+		PreferenceManager preferencesManager = JavaLanguageServerPlugin.getPreferencesManager();
+		if (projectsManager != null && preferencesManager != null) {
+			List<String> oldSha256Allowed = oldPreferences.getSha256Allowed() == null ? Collections.emptyList() : oldPreferences.getSha256Allowed();
+			List<String> newSha256Allowed = newPreferences.getSha256Allowed() == null ? Collections.emptyList() : newPreferences.getSha256Allowed();
+			List<String> oldSha256Disallowed = oldPreferences.getSha256Disallowed() == null ? Collections.emptyList() : oldPreferences.getSha256Disallowed();
+			List<String> newSha256Disallowed = newPreferences.getSha256Disallowed() == null ? Collections.emptyList() : newPreferences.getSha256Disallowed();
+			if ((!Objects.equals(oldSha256Allowed, newSha256Allowed) || !Objects.equals(oldSha256Disallowed, newSha256Disallowed))) {
+				for (IProject project : ProjectUtils.getGradleProjects()) {
+					String projectDir = project.getLocation().toFile().getAbsolutePath();
+					Path projectPath = Paths.get(projectDir);
+					if (preferencesManager != null && preferencesManager.getPreferences().isGradleWrapperEnabled() && Files.exists(projectPath.resolve("gradlew"))) {
+						ProjectConfiguration configuration = CorePlugin.configurationManager().loadProjectConfiguration(project);
+						GradleDistribution distribution = configuration.getBuildConfiguration().getGradleDistribution();
+						if (!(distribution instanceof WrapperGradleDistribution)) {
```

Comment:
why updating a project that doesn't use the wrapper when you're checking wrapper checksum changes?

Reply:
It is required if you update the wrapper's checksum from false to true.
You can try the gradle-4.0 project:
- add 
```
"java.imports.gradle.wrapper.checksums": [
        {
            "sha256": "41c8aa7a337a44af18d8cda0d632ebba469aef34f3041827624ef5c1a4e4419d",
            "allowed": false
        }
    ],
```
- check .settings/org.eclipse.buildship.core.prefs. You will see that we use 
```
connection.gradle.distribution=GRADLE_DISTRIBUTION(VERSION(6.3))
```
- change the property:
"java.imports.gradle.wrapper.checksums": [
        {
            "sha256": "41c8aa7a337a44af18d8cda0d632ebba469aef34f3041827624ef5c1a4e4419d",
            "allowed": true
        }
    ],
```
- check .settings/org.eclipse.buildship.core.prefs.
```
connection.gradle.distribution=GRADLE_DISTRIBUTION(WRAPPER)
```